### ISSUE TITLE
[BUG] Protocol not available (src/tcp_connecter.cpp:290)

### ISSUE THREAD

**Issue by morfien101** [saltstack/salt#62495]
**Description**
Following the authors instructions in 'Salt Open: Automating Your Enterprise and Your Network', the CentOS 8 Minion can not connect to the master. This is not due to a configuration error but rather the minion crashes with a stack trace being produced. I have done some troubleshooting to try fix this but to no avail. I am not sure what is happening so I'm not sure what to try fix.

I worry that someone else tries this book and gets stuck in the same place. As its right at the beginning of the book it kinda hurts the entry to the product.

Started via systemctl
```text
[root@minion1 ~]# journalctl -u salt-minion.service 
-- Logs begin at Sun 2022-08-21 10:01:08 SAST, end at Sun 2022-08-21 10:06:19 SAST. --
Aug 21 10:03:14 minion1 systemd[1]: Starting The Salt Minion...
Aug 21 10:03:14 minion1 systemd[1]: Started The Salt Minion.
Aug 21 10:03:15 minion1 salt-minion[24322]: Protocol not available (src/tcp_connecter.cpp:290)
Aug 21 10:03:15 minion1 salt-minion[24322]: #0  0xffcddaef6de0 in ? (?+0x7fe6e6b675f0)
Aug 21 10:03:15 minion1 salt-minion[24322]: #1  0xffcddaf35ea0 in ? (?+0x7fe6e6b675f0)
Aug 21 10:03:15 minion1 salt-minion[24322]: #2  0xffcddaf36010 in ? (?+0x7fe6e6b675f0)
Aug 21 10:03:15 minion1 salt-minion[24322]: #3  0xffcddaef6070 in ? (?+0x7fe6e6b675f0)
Aug 21 10:03:15 minion1 salt-minion[24322]: #4  0xffcddaf37180 in ? (?+0x7fe6e6b675f0)
Aug 21 10:03:15 minion1 salt-minion[24322]: #5  0x7fe6fdd9214a in /lib64/libpthread.so.0 (start_thread+0xea)
Aug 21 10:03:15 minion1 salt-minion[24322]: #6  0x7fe6fd337df0 in /lib64/libc.so.6 (clone+0x43)
Aug 21 10:03:15 minion1 systemd[1]: salt-minion.service: Main process exited, code=exited, status=250/n/a
Aug 21 10:03:15 minion1 systemd[1]: salt-minion.service: Failed with result 'exit-code'.
```
Debug log enabled
```text
sudo salt-minion -l debug
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/99-master-address.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/99-master-address.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: minion1
[DEBUG   ] Using importlib_metadata to load entry points
[DEBUG   ] Override  __grains__: <module 'salt.loaded.int.log_handlers.sentry_mod' from '/usr/lib/python3.6/site-packages/salt/log/handlers/sentry_mod.py'>
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[INFO    ] Setting up the Salt Minion "minion1"
[DEBUG   ] Marking 'base64_encode' as a jinja filter
[DEBUG   ] Marking 'base64_decode' as a jinja filter
[DEBUG   ] Marking 'md5' as a jinja filter
[DEBUG   ] Marking 'sha1' as a jinja filter
[DEBUG   ] Marking 'sha256' as a jinja filter
[DEBUG   ] Marking 'sha512' as a jinja filter
[DEBUG   ] Marking 'hmac' as a jinja filter
[DEBUG   ] Marking 'hmac_compute' as a jinja filter
[DEBUG   ] Marking 'random_hash' as a jinja filter
[DEBUG   ] Marking 'rand_str' as a jinja filter
[DEBUG   ] Marking 'file_hashsum' as a jinja filter
[DEBUG   ] Marking 'http_query' as a jinja filter
[DEBUG   ] Marking 'strftime' as a jinja filter
[DEBUG   ] Marking 'date_format' as a jinja filter
[DEBUG   ] Marking 'raise' as a jinja global
[DEBUG   ] Marking 'match' as a jinja test
[DEBUG   ] Marking 'equalto' as a jinja test
[DEBUG   ] Marking 'skip' as a jinja filter
[DEBUG   ] Marking 'sequence' as a jinja filter
[DEBUG   ] Marking 'to_bool' as a jinja filter
[DEBUG   ] Marking 'indent' as a jinja filter
[DEBUG   ] Marking 'tojson' as a jinja filter
[DEBUG   ] Marking 'quote' as a jinja filter
[DEBUG   ] Marking 'regex_escape' as a jinja filter
[DEBUG   ] Marking 'regex_search' as a jinja filter
[DEBUG   ] Marking 'regex_match' as a jinja filter
[DEBUG   ] Marking 'regex_replace' as a jinja filter
[DEBUG   ] Marking 'uuid' as a jinja filter
[DEBUG   ] Marking 'unique' as a jinja filter
[DEBUG   ] Marking 'min' as a jinja filter
[DEBUG   ] Marking 'max' as a jinja filter
[DEBUG   ] Marking 'avg' as a jinja filter
[DEBUG   ] Marking 'union' as a jinja filter
[DEBUG   ] Marking 'intersect' as a jinja filter
[DEBUG   ] Marking 'difference' as a jinja filter
[DEBUG   ] Marking 'symmetric_difference' as a jinja filter
[DEBUG   ] Marking 'method_call' as a jinja filter
[DEBUG   ] Marking 'yaml_dquote' as a jinja filter
[DEBUG   ] Marking 'yaml_squote' as a jinja filter
[DEBUG   ] Marking 'yaml_encode' as a jinja filter
[DEBUG   ] Created pidfile: /var/run/salt-minion.pid
[INFO    ] Starting up the Salt Minion
[DEBUG   ] AsyncEventPublisher PUB socket URI: /var/run/salt/minion/minion_event_1e9498c860_pub.ipc
[DEBUG   ] AsyncEventPublisher PULL socket URI: /var/run/salt/minion/minion_event_1e9498c860_pull.ipc
[INFO    ] Starting pull socket on /var/run/salt/minion/minion_event_1e9498c860_pull.ipc
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_1e9498c860_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_1e9498c860_pull.ipc
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/99-master-address.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/99-master-address.conf
[DEBUG   ] Override  __utils__: <module 'salt.loaded.int.grains.zfs' from '/usr/lib/python3.6/site-packages/salt/grains/zfs.py'>
[DEBUG   ] Elapsed time getting FQDNs: 0.11085319519042969 seconds
[DEBUG   ] LazyLoaded zfs.is_supported
[INFO    ] Creating minion process manager
[DEBUG   ] Process Manager starting!
[DEBUG   ] Process Manager starting!
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Master URI: tcp://192.168.56.100:4506
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'minion1', 'tcp://192.168.56.100:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (1361)
[DEBUG   ] Setting zmq_reconnect_ivl to '1361ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.56.100:4506
[DEBUG   ] Trying to connect to: tcp://192.168.56.100:4506
[DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
Protocol not available (src/tcp_connecter.cpp:290)
#0  0xffc2db1c9de0 in ? (?+0x7fe166a535f0)
#1  0xffc2db208ea0 in ? (?+0x7fe166a535f0)
#2  0xffc2db209010 in ? (?+0x7fe166a535f0)
#3  0xffc2db1c9070 in ? (?+0x7fe166a535f0)
#4  0xffc2db20a180 in ? (?+0x7fe166a535f0)
#5  0x7fe17d48814a in /lib64/libpthread.so.0 (start_thread+0xea)
#6  0x7fe17ca2ddf0 in /lib64/libc.so.6 (clone+0x43)
```

``` text
[root@minion1 ~]# uname -a
Linux minion1 4.18.0-277.el8.x86_64 #1 SMP Wed Feb 3 20:35:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[root@minion1 ~]# cat /etc/redhat-release 
CentOS Stream release 8
```

This Vagrantfile will help reproduce the problem

```ruby
Vagrant.configure("2") do |config|
  config.vm.define "master" do |master|
    master.vm.box = "ubuntu/focal64"
    master.vm.network "private_network", ip: "192.168.56.100"
    master.vm.hostname = "master"
    master.vm.provider "virtualbox" do |vb|
      vb.memory = "1024"
    end
    master.vm.provision "shell", path: "scripts/common.sh"
    master.vm.provision "shell", path: "scripts/master_install.sh"
  end
  
  config.vm.define "minion1" do |minion1|
    minion1.vm.box = "centos/stream8"
    minion1.vm.network "private_network", ip: "192.168.56.101"
    minion1.vm.hostname = "minion1"
    minion1.vm.provider "virtualbox" do |vb|
      vb.memory = "1024"
    end
    minion1.vm.provision "shell", inline: "curl -o install_salt.sh -L https://bootstrap.saltproject.io"
    minion1.vm.provision "shell", inline: "sudo sh install_salt.sh -i minion1 -A 192.168.56.100 stable 3003"
  end
end
```

Commands to see the problem
```sh
vagrant up master minion1
vagrant ssh minion1

# In minion1
sudo systemctl status salt-minion.service
sudo salt-minion -l debug
```

I'm not sure what is to be done about this, if its a legitimate problem/bug. However, I couldn't find anything that helped me fix it. Even if it is a easy fix something that shows up in a google search would be much nicer than the experience that I had.

**Comment by morfien101** [comment]
Link to the book: https://link.springer.com/book/10.1007/978-1-4842-7237-4

**Comment by morfien101** [comment]
I have tried this on a different CentOS 8 box with a newer kernel and a CentOS7 box just to see if its isolated to CentOS8. Spoiler, it's not.

```ruby
minion1.vm.box = "generic/centos8"
```
```text
[vagrant@minion1 ~]$ uname -a
Linux minion1 4.18.0-348.7.1.el8_5.x86_64 #1 SMP Wed Dec 22 13:25:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
[vagrant@minion1 ~]$ cat /etc/redhat-release 
CentOS Linux release 8.5.2111
```

```ruby
minion1.vm.box = "generic/centos7"
```
```text
[root@minion1 ~]# openssl version
OpenSSL 1.0.2k-fips  26 Jan 2017
[root@minion1 ~]# uname -a
Linux minion1 3.10.0-1160.71.1.el7.x86_64 #1 SMP Tue Jun 28 15:37:28 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
[root@minion1 ~]# cat /etc/redhat-release 
CentOS Linux release 7.9.2009 (Core)
```

**Comment by OrangeDog** [comment]
What exactly does the book tell you to do? You appear to have configured TCP transport instead of the default ZMQ?

Have you tried with the latest release (3004.2)? Can you give the output of `salt-call --local --versions`?

**Comment by OrangeDog** [comment]
The error appears to be coming from this assertion: https://github.com/zeromq/libzmq/blob/v4.3.4/src/tcp_connecter.cpp#L290

Have you disabled UDP in your vagrant network or similar?

**Comment by morfien101** [comment]
@OrangeDog I have managed to narrow down who is causing the issue but not what is causing it.

TL;DR answer:
The issue appears to be the network adaptor of VirtualBox when using the Host-Only setting. This is what is created when you use the "private_network" setting in vagrant. However it appears it is only an issue when I use the vboxnet0 interface. Adding a new network and deleting vboxnet0 appears to solve the issue.

More details:
I use the clue that you gave to me about UDP networks. I can't think of a way that I would have inadvertently disabled UDP. It's at the start of the book and we have done nothing but download an image (several different ones) and installed salt. That got me thinking about vagrant. I know it does things on the machine when it provisions it. However disabling UDP seem really strange. 

After much reading I tried the minion2 instance which is Ubuntu. It also has an issue.

In the book we setup the master as a master and minion. The master works fine, you can connect, test.ping and everything.

 I then for the hell of it tried changing the network interface to see if it was a hardware issue (be it virtual) and after setting the second 'private' interface to NAT, salt started. It could not connect to the master for obvious reasons but the service started without crashing.
After various other fiddling I changed the interface back to Host-Only and it failed as above again.
I then decided to try a different internal network. Change the minion1 instance to it and it magically started without crashing. I then changed the master to the newly created instance and it also started AND connected with the minion.

I have no idea what is going on under the hood but its now working.
Both of the below OS were affected
Linux (Linux pop-os 5.17.15-051715-generic). Virtualbox 6.1.34 r150636
Apple Macbook Pro MacOS 12.5. Virtualbox 6.1.36 R152435

As a side note, upgrading to the latest version of salt did not help the issue.

I really hope that someone else that has this issue will find this post. I'd be ok with you closing this issue if you see no further benefit pursuing this issue as it doesn't seem salt related and rather VirtualBox related.
